Iframes auth redirect

sameerag opened the pull request:
This PR in conjunction with #939 provides an option for applications living out of an iframe to handle interactive requests using "redirect".

**Problem description:** 
Some identity flows cannot be completed in an iframe. In particular:

- Credential Entry
- Consent

Clickjacking  is an attack whereby a malicious site iframes a trusted site and displays it to the user while simultaneously providing a transparent iframe on top of the trusted site. This transparent iframe is invisible to the user, but is able to record keystrokes (e.g. the password). Alternatively, a malicious site can move the trusted iframes it shows, permitting them to always move the "Consent" button to directly under the users mouse. Without going into detail, the only mitigation here is to forbid iframes on these flows. Identity permits services to use iframes only in combination with prompt=none, which ensures that these flows are not hit.

These limitations placed by Identity Service block any user interaction in an iframe app. This effectively requires all applications to be able to use a top-frame (either the current frame or a pop-up frame) to solve user interaction required prompts.

To mitigate this and have applications perform authentication in iframes, we are adding the below changes for different flows:

- [x] #939 helps with silent and popup calls where additional logic is added to monitor the popup/iframe opened by the MSAL JS to process the token as soon as the service redirects the request. APIs that are supported in iframed applications with this change are:
 - loginPopup()
- acquireTokenPopup()
 - acquireTokenSilent()

- [x] #975 (this PR) supports the "redirect" APIs by establishing a trusted relationship with the parent application (the top frame). As expanded above, the iframed application is prohibited from any interaction by microsoft identity service and hence this can be supported only with a trusted relationship with the parent application (which is a full window and hence can render interactive screens)

Detailed documentation on the protocol applied here will be published with the release of both these PRs.

sameerag commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -256,7 +238,55 @@ export class UserAgentApplication {
+            if(WindowUtils.isWindowOnTop() && this.cacheStorage.getItem("topFrameURI")) { // replace with a cache check for onbehalf that is more apt if needed
+                this.cacheStorage.setItem("iframedAppHash", urlHash);
+                this.cacheStorage.removeItem("topFrameURI");
+                this.navigateWindow(this.cacheStorage.getItem("topFrameURI"));
```
Don't clear the cache before retrieving the URL

sameerag replied:
Done

sameerag commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -863,7 +892,16 @@ export class UserAgentApplication {
+            if(WindowUtils.isInIframe()) {
+                // posting up to redirect on the iframed app's behalf
+                this.cacheStorage.setItem("urlNavigate", urlNavigate);
+                window.top.postMessage("{redirectRequest: yes}", window.location.href);
```
add message abstraction for all post messages

sameerag commented on lib/msal-core/src/utils/UrlUtils.ts:
```diff
@@ -97,7 +97,7 @@ export class UrlUtils {
      * Returns current window URL as redirect uri
      */
-    static getDefaultRedirectUri(): string {
+    static getDefaultUri(): string {
```
Done

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+export type MESSAGE_SCHEMA = {
+    type: message_type,
+    content?: message_content,
+    data?: string
```
Done

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+
+export type MESSAGE_SCHEMA = {
+    type: message_type,
+    content?: message_content,
```
Merged both as discussed and made it:

PAYLOAD = {
    type: MessageType,
    data?: data
}

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+import { MessageDispatcher } from "./MessageDispatcher";
+
+export enum message_content {
+    REDIRECT_REQUEST = "REDIRECT_REQUEST",
```
resolved below

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+export enum message_type {
+    REDIRECT_REQUEST = "REDIRECT_REQUEST",
+    URL = "URL",
+    HASH = "HASH"
```
resolved below

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+    HASH = "HASH"
+};
+
+export enum window_type {
```
Done

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+    URL_NAVIGATE = "URL_NAVIGATE"
+};
+
+export enum message_type {
```
Done

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+import { Logger } from "./../Logger";
+import { MessageDispatcher } from "./MessageDispatcher";
+
+export enum message_content {
```
Removed this

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,103 @@
+    data?: string
+};
+
+export class MessageHelper {
```
It is more of a helper as all constants and utils are combined in this.

sameerag commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -233,9 +244,23 @@ export class UserAgentApplication {
         const urlContainsHash = UrlUtils.urlContainsHash(urlHash);
 
+        // read the hash stored through the topframe in redirect by delegation flow
+        const urlTopFrame = this.messageCache.read(message_content.URL_TOP_FRAME);
```
I think UrlTopFrame is more appropriate, it is literally the URL of the top frame

jasonnutter commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,95 @@
+    IFRAME
+};
+
+export type PAYLOAD = {
```
Does this need to be all caps?

sameerag replied:
Need not be. Changed.

sameerag replied:
Also I guess Daryl's initial suggestion was to name the data payload, not the entire message. Reflecting that comment too.

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,86 @@
+            case WindowType.IFRAME: {
+
+                // check the origin, should match window.top always; message channel may be more secure
+                if (window.top != event.source) {
```
`!==`

sameerag replied:
+1

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,86 @@
+     * This will be a unique handler per message, we will allow only one active request at a time
+     * @param event
+     */
+    private receiveMessage(event: any) {
```
Is there not a specific type you can use?

sameerag replied:
I tried MessageEvent but it did not seem to work. I did not find anything appropriate except `Object` or `any`.

jasonnutter replied:
Hmm why didnt `MessageEvent` work?

pkanher617 replied:
Interested to know why message event didn't work, seems like we lose some type checking there...

sameerag replied:
Message dispatching with event.source which should be a `WindowProxy` is not type matching. I will dig into this later with a separate PR.

jasonnutter commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,95 @@
+        if(WindowUtils.isWindowOnTop()) {
+            return WindowType.TOP_FRAME;
+        }
+        else if(WindowUtils.isInIframe()) {
```
Don't need `else` since previous blocks are returning.

sameerag replied:
Of course. Done.

jasonnutter commented on lib/msal-core/src/Configuration.ts:
```diff
@@ -95,6 +96,17 @@ export type FrameworkOptions = {
+ * - topFrameOrigin             - origin check to restrict messages to the top frame origin only
+ * - consentNeeded              - indicates if the library needs consent from the topframe to delegate interaction from the embedded (iframe) application
+ */
+export type BrokerOptions = {
```
Could this be an enum?

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    public setCallBack(consentNeeded: boolean, callback: iframeRedirectCallback) {
+        this.consentNeeded = consentNeeded;
```
I dont think you actually need this boolean. The truthiess of `iframeRedirectCallback` will tell you if you should invoke it.

sameerag replied:
Done

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,113 @@
+     * top frame application calls this function to ack to proceed
+     * @param message
+     */
+    processIframeRedirectCallback(url: string) {
```
A return type here I think would make it clear that this is returning a new function.

jasonnutter replied:
Another thing that we should just enforce with a lint rule.

sameerag replied:
Done

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,113 @@
+
+                    case MessageType.URL_NAVIGATE: {
+                        // if the response is the URL to navigate for token acquisition, navigate to STS
+                        if (receivedMessage.type === MessageType.URL_NAVIGATE) {
```
This check is superfluous as it will always be true since this is switch statement.

sameerag replied:
Sorry yes! I changed the old code and forgot remove the obsolete check. Thanks.

jasonnutter commented on lib/msal-core/src/messaging/MessageListener.ts:
```diff
@@ -0,0 +1,113 @@
+                    return;
+                }
+
+                if (receivedMessage.type === MessageType.URL_TOP_FRAME) {
```
I might argue a nest switch statement would be good for consistency with the above block.

sameerag replied:
Done

sameerag commented on lib/msal-core/src/utils/WindowUtils.ts:
```diff
@@ -201,4 +212,23 @@ export class WindowUtils {
+     */
+    static navigateWindow(urlNavigate: string, logger: Logger, popupWindow?: Window) {
+        // Navigate if valid URL
+        if (urlNavigate && !StringUtils.isEmpty(urlNavigate)) {
```
This was legacy code but I agree with you. Done.

pkanher617 commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -223,9 +249,23 @@ export class UserAgentApplication {
         // On the server 302 - Redirect, handle this
         if (!this.config.framework.isAngular && urlContainsHash && !WindowUtils.isInIframe() && !WindowUtils.isInPopup()) {
-            this.handleAuthenticationResponse(urlHash);
+            // REDIRECT_IFRAMES: if we are in topframe, store the hash in the cache
```
This feels like this block should probably be in a helper.

sameerag replied:
Done

pkanher617 commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -516,8 +566,14 @@ export class UserAgentApplication {
 
-            // prompt user for interaction
-            this.navigateWindow(urlNavigate, popUpWindow);
+            // IFRAMEDAPPS: if we are redirecting in an iframe, post a message to the topFrame
```
Put this in helper.

sameerag replied:
I don't think it adds any value here except too much variable passing.

sameerag commented on lib/msal-core/src/messaging/MessageHelper.ts:
```diff
@@ -0,0 +1,95 @@
+    REDIRECT_REQUEST = "REDIRECT_REQUEST",
+    HASH = "HASH",
+    URL_TOP_FRAME = "URL_TOP_FRAME",
+    URL_NAVIGATE = "URL_NAVIGATE"
```
This is to pass strings as keys in the MessageCache. We can change it but it is the model we followed for `Constants` across for now. May be a future `coding standards` project?

sameerag commented:
@pkanher617 @jasonnutter Addressed all the comments. Please review. I have retained some open issues on naming/type checks, will address them separately post release.
